File: nodejs/src/DICOM2FHIRImagingStudy.js
const fs = require('fs');
const dicomParser = require('dicom-parser');
const Moment = require('moment');
const _ = require('lodash');
const flatten = require('flat');
const { ImagingStudy, ImagingStudy_Series, ImagingStudy_Series_Instance, Coding, Identifier } = require("./FHIRDataTypes");

async function doDICOMToFHIRImagingStudy(filename) {
    return new Promise(async (resolve) => {
        try {
            let dataset = dicomParser.parseDicom(filename , {
                untilTag: "x7fe00010"
            });
            let studyObj = new ImagingStudy();
            let studyInstanceUID = dataset.string('x0020000d');
            let ANandIssuer = await getTwoTag(dataset, 'x00080050', 'x00080051'); //Accession Number and Issuer
            studyObj.id = studyInstanceUID;
            let identifiers = [studyInstanceUID, ANandIssuer, dataset.string('x00200010')];
            studyObj.identifier = await getImagingStudyIdentifiers(identifiers);
            studyObj.modality = dataset.string('x00080061');
            let patientId = dataset.string('x00100020');
            if (patientId) {
                studyObj.subject.reference = "Patient/" + dataset.string('x00100020');
                studyObj.subject.type = "Patient";
                studyObj.subject.identifier.use = "usual"
                studyObj.subject.identifier.value = dataset.string('x00100020');
            } else {
                studyObj.subject.reference = "Patient/unknown"
                studyObj.subject.type = "Patient";
                studyObj.subject.identifier.use = "anonymous"
                studyObj.subject.identifier.value = "unknown";
            }

            let imaging_started = dataset.string('x00080020') + dataset.string('x00080030');
            const date = Moment(imaging_started, "YYYYMMDDhhmmss").toISOString();
            studyObj.started = date;
            studyObj.numberOfSeries = dataset.string('x00201206');
            studyObj.numberOfInstances = dataset.string('x00201208');
            studyObj.description = dataset.string('x00081030');
            let studySeriesObj = new ImagingStudy_Series();
            studySeriesObj.uid = dataset.string('x0020000e');
            studySeriesObj.number = dataset.intString('x00200011');
            studySeriesObj.modality.code = dataset.string('x00080060');
            studySeriesObj.description = dataset.string('x0008103e');
            studySeriesObj.numberOfInstances = dataset.intString('x00201209');
            studySeriesObj.bodySite.display = dataset.string('x00180015');
            let series_started = dataset.string('x00080021') + dataset.string('x00080031');
            const series_date = Moment(series_started, "YYYYMMDDhhmmss").toDate();
            studySeriesObj.started = series_date != null ? series_date : undefined;
            studySeriesObj.performer = dataset.string('x00081050') || dataset.string('x00081052') || dataset.string('x00081070') || dataset.string('x00081072');
            let studySeriesInstanceObj = new ImagingStudy_Series_Instance();
            studySeriesInstanceObj.uid = dataset.string('x00080018');
            studySeriesInstanceObj.sopClass.system = "urn:ietf:rfc:3986"
            studySeriesInstanceObj.sopClass.code = "urn:oid:" + dataset.string('x00080016');
            studySeriesInstanceObj.number = dataset.intString('x00200013');
            studySeriesInstanceObj.title = dataset.string('x00080008') || dataset.string('x00070080') || ((dataset.string('x0040a043') != undefined) ? dataset.string('x0040a043') + dataset.string('x00080104') : undefined) || dataset.string('x00420010');
            let imagingStudyJson = await CombineImagingStudyClass(studyObj, studySeriesObj, studySeriesInstanceObj);
            resolve(imagingStudyJson);
        }
        catch (ex) {
            console.log(ex);
            resolve(false);
        }
    });
}

async function getTwoTag(dataset, I_Tag1, I_Tag2) {
    return new Promise((resolve) => {
        let str1 = dataset.string(I_Tag1);
        let str2 = dataset.string(I_Tag2);
        let result = "";
        if (str1 != undefined && str2 != undefined) {
            result = str1 + str2;
        }
        else if (str1 != undefined) {
            result = str1;
        }
        else if (str2 != undefined) {
            result = str2;
        }
        else {
            result = undefined;
        }
        return resolve(result);
    });
}
//Common just use official
async function getImagingStudyIdentifiers(identifiers) {
    return new Promise((resolve) => {
        let result = [];
        // StudyInstanceUID
        if (identifiers[0] != undefined) {
            let identifier1 = new Identifier();
            identifier1.use = "official";
            identifier1.system = "urn:dicom:uid";
            identifier1.value = "urn:oid:" + identifiers[0];
            result.push(identifier1);
        }
        //need sample dicom with the organization
        //Accession Number and Issuer
        if (identifiers[1] != undefined) {
            let identifier2 = new Identifier();
            identifier2.type = new Coding();
            identifier2.use = "usual";
            identifier2.value = identifiers[1];
            result.push(identifier2);
        }
        //Study ID
        if (identifiers[2] != undefined) {
            let identifier3 = new Identifier();
            identifier3.use = "secondary";
            identifier3.value = "s" + identifiers[2];
            result.push(identifier3);
        }
        return resolve(result);
    });
}

//http://jsfiddle.net/ryeballar/n0afoxdu/
//用於清除Object裡面undefined、空Object以及空Array
function removeEmpty(obj) {
    if (_.isArray(obj)) {
        return _(obj)
            .filter(_.isObject)
            .map(removeEmpty)
            .reject(_.isEmpty)
            .concat(_.reject(obj, _.isObject))
            .value();
    }
    return _(obj)
        .pickBy(_.isObject)
        .mapValues(removeEmpty)
        .omitBy(_.isEmpty)
        .assign(_.pickBy(_.omitBy(obj, _.isObject) , _.identity))
        .value();
}
/**
 * 將Study, Series, Instance組合成一個ImagingStudy
 * @param {*} imagingStudy 
 * @param {*} imagingStudySeries 
 * @param {*} imagingStudySeriesInstance 
 */
async function CombineImagingStudyClass(imagingStudy, imagingStudySeries, imagingStudySeriesInstance) {
    try {
        let imagingStudyJson = imagingStudy.ToJson();
        let imagingStudySeriesJson = imagingStudySeries.ToJson();
        let imagingStudySeriesInstanceJson = imagingStudySeriesInstance.ToJson();
        imagingStudySeriesJson.instance.push(imagingStudySeriesInstanceJson);
        imagingStudyJson.series.push(imagingStudySeriesJson);
        let ImagingStudyJsonFlatten = flatten(imagingStudyJson);
        ImagingStudyJsonFlatten = removeEmpty(ImagingStudyJsonFlatten);
        imagingStudyJson = flatten.unflatten(ImagingStudyJsonFlatten);
        return imagingStudyJson;
    } catch (e) {
        console.error(e);
        return false;
    }
}
async function getFileDicomParser(filename) {
    let isExist = fs.existsSync(filename);
    if (isExist) {
        let file = fs.readFileSync(filename);
        return file;
    } else {
        if (_.isString(filename)) {
            return false;
        }
        return filename;
    }
}
module.exports.DCM2FHIR = async (filename) => {
    return new Promise(async (resolve) => {
        try {
            let dicomfile = "";
            //取得local file 或 api 上的blob file
            dicomfile = await getFileDicomParser(filename);
            let imagingStudyJson = await doDICOMToFHIRImagingStudy(dicomfile);
            return resolve({
                status: true,
                message: "processing successfully",
                data: imagingStudyJson
            });
        } catch (err) {
            console.error(err);
            return resolve({
                status: false ,
                message: "processing DICOM to FHIR ImagingStudy error",
                data: err
            });
        }
    });
}

File: python/DICOM2FHIRImagingStudy.py
import pydicom
import json
import utils
from datetime import datetime


class Coding:
    def __init__(self):
        self.system = None
        self.version = None
        self.code = None
        self.display = None
        self.userSelected = None

    pass


pass


class CodeableConcept:
    def __init__(self):
        self.Coding = None
        self.text = None

    pass


pass


class Identifier:
    def __init__(self):
        self.use = None
        self.type = None
        self.system = None
        self.value = None
        self.period = None

    pass


pass


class Period:
    def __init__(self):
        self.start = None
        self.end = None

    pass


pass


class Reference:
    def __init__(self):
        self.reference = None
        self.type = None
        self.identifier = None
        self.display = None

    pass


pass


class ImagingStudy:
    def __init__(self):
        self.resourceType = "ImagingStudy"
        self.id = None
        self.identifier = None
        self.status = "unknown"
        self.modality = None
        self.subject = None
        self.started = None
        self.endpoint = None
        self.numberOfSeries = None
        self.numberOfInstances = None
        self.description = None
        self.series = None

    pass


pass


class ImagingStudySeries:
    def __init__(self):
        self.uid = None
        self.number = None
        self.modality = Coding()
        self.modality.system = "http://dicom.nema.org/resources/ontology/DCM"
        self.description = None
        self.numberOfInstances = None
        self.endpoint = None
        self.bodySite = None
        self.laterality = None
        self.started = None
        self.performer = None
        self.instance = None

    pass


pass


class ImagingStudySeriesInstance:
    def __init__(self):
        self.uid = None
        self.sopClass = Coding()
        self.number = None
        self.title = None

    pass


pass


def get_accession_number_and_issuer(i_data_set):
    accession_number = i_data_set.get(0x00080050)
    issuer = i_data_set.get(0x00080051)
    result = None
    if accession_number is not None and issuer is not None:
        result = accession_number.value + issuer.value
    elif accession_number is not None:
        result = accession_number.value
    elif issuer is not None:
        result = issuer.value
    pass
    return result


pass


def get_imaging_study_identifier_list(i_data_set):
    result = []
    study_instance_uid = i_data_set[0x0020000D].value
    accession_number_and_issuer = get_accession_number_and_issuer(i_data_set)
    study_id = i_data_set.get(0x00200010)
    identifier_study_instance_uid = Identifier()
    identifier_study_instance_uid.use = "official"
    identifier_study_instance_uid.system = "urn:dicom:oid"
    identifier_study_instance_uid.value = "urn:oid" + study_instance_uid
    result.append(identifier_study_instance_uid)
    if accession_number_and_issuer is not None:
        identifier_accession_number_and_issuer = Identifier()
        identifier_accession_number_and_issuer.use = "usual"
        identifier_accession_number_and_issuer.value = accession_number_and_issuer
        result.append(identifier_accession_number_and_issuer)
    pass
    if study_id is not None:
        identifier_study_id = Identifier()
        identifier_study_id.use = "secondary"
        identifier_study_id.value = study_id.value
        result.append(identifier_study_id)
    pass
    return result


pass


def get_imaging_started(date, time):
    imaging_started = None
    imaging_started_str = None
    if date is not None and time is not None:
        imaging_started = date.value + time.value
    elif date is not None:
        imaging_started = date.value
    elif date is not None:
        imaging_started = date.value
    pass
    format_data = "%Y%M%d%H%M%S"
    if imaging_started is not None:
        imaging_started_str = datetime.strptime(imaging_started, format_data)
    pass
    return imaging_started_str


pass


def do_dicom_to_fhir_imagingstudy(filename):
    data_set = pydicom.dcmread(filename, force=True, stop_before_pixels=True)
    imaging_study_obj = ImagingStudy()
    study_instance_uid = data_set[0x0020000D].value
    imaging_study_obj.id = study_instance_uid
    imaging_study_obj.identifier = get_imaging_study_identifier_list(data_set)
    patient_id = data_set.get(0x00100020)

    imaging_study_obj.subject = Reference()
    imaging_study_obj.subject.identifier = Identifier()

    if patient_id is not None:
        imaging_study_obj.subject.reference = "Patient/" + patient_id.value
        imaging_study_obj.subject.type = "Patient"
        imaging_study_obj.subject.identifier.use = "usual"
        imaging_study_obj.subject.identifier.value = patient_id.value
    else:
        imaging_study_obj.subject.reference = "Patient/unknown"
        imaging_study_obj.subject.type = "type"
        imaging_study_obj.subject.identifier.use = "anonymous"
        imaging_study_obj.subject.identifier.value = "unknown"
    pass
    imaging_study_obj.started = get_imaging_started(data_set.get(0x00800020), data_set.get(0x00800030))

    number_of_series = data_set.get(0x00201206)
    if number_of_series is not None:
        imaging_study_obj.numberOfSeries = number_of_series.value
    pass

    number_of_study_instances = data_set.get(0x00201208)
    if number_of_study_instances is not None:
        imaging_study_obj.numberOfInstances = number_of_study_instances.value
    pass

    study_description = data_set.get(0x00081030)
    if study_description is not None:
        imaging_study_obj.description = study_description.value
    pass

    series_obj = ImagingStudySeries()
    series_obj.uid = data_set[0x0020000e].value

    series_number = data_set.get(0x00200011)
    if series_number is not None:
        series_obj.number = int(series_number.value)
    pass

    modality_code = data_set.get(0x00080060)
    if modality_code is not None:
        series_obj.modality = Coding()
        series_obj.modality.code = modality_code.value
    pass

    series_description = data_set.get(0x0008103e)
    if series_description is not None:
        series_obj.description = series_description.value
    pass

    number_of_series_instances = data_set.get(0x00201209)
    if number_of_series_instances is not None:
        series_obj.numberOfInstances = number_of_series_instances.value
    pass

    series_body_site = data_set.get(0x00180015)
    if series_body_site is not None:
        series_obj.bodySite = Coding()
        series_obj.bodySite = series_body_site.value
    pass

    series_obj.started = get_imaging_started(data_set.get(0x00080021), data_set.get(0x00080031))

    performer = data_set.get(0x00081050) or data_set.get(0x00081052) or data_set.get(0x00081070) or \
                data_set.get(0x00081072)
    if performer is not None:
        series_obj.performer = performer.value
    pass

    instance_obj = ImagingStudySeriesInstance()
    instance_obj.uid = data_set[0x00080018].value
    instance_obj.sopClass = Coding()
    instance_obj.sopClass.system = "urn:ietf:rfc:3986"
    instance_obj.sopClass.code = "urn:oid" + data_set[0x00080016].value

    instance_number = data_set.get(0x00200013)
    if instance_number is not None:
        instance_obj.number = instance_number.value
    pass

    series_obj.instance = [instance_obj]
    imaging_study_obj.series = [series_obj]
    imaging_study_obj_json_str = json.dumps(imaging_study_obj, default=lambda o: o.__dict__)
    imaging_study_obj_json = json.loads(imaging_study_obj_json_str)
    imaging_study_obj_json_remove = utils.delete_none(imaging_study_obj_json)
    imaging_study_obj_json_str = json.dumps(imaging_study_obj_json_remove, default=lambda o: o.__dict__, indent=4)
    print(imaging_study_obj_json_str)
    return imaging_study_obj_json_str


pass


if __name__ == '__main__':
    do_dicom_to_fhir_imagingstudy("../image-000001.dcm")

pass
